src/core/qgsrendercontext.h

```diff
@@ -723,6 +723,20 @@ class CORE_EXPORT QgsRenderContext
      */
     bool isGuiPreview() const { return mIsGuiPreview; }
 
+    /**
+     * Set arbitrary custom flags which consumers might honour to adapt their rendering.
+     * \param flags List of string flags
+     * \since QGIS 3.12
+     */
+    void setCustomRenderFlags( const QStringList &flags ) { mCustomRenderFlags = flags; }
```

Comment:
That would not really be compatible with `QgsMapSettings::customRenderFlags` returning a string of `;`-separated values though? The `QStringList` is just convenience in that the users can use `contains` directly without having to `;`-split the string before, but for the rest the contents will equal to `QgsMapSettings::customRenderFlags`.

Reply:
What kind of thing are you actually passing here? If it's a variant map you could store bools easily (more easily than parsing a string list!)

Reply:
See for instance [1] and [2]. Thing is that `QgsMapSettings::customRenderFlags` is pre-existing, so to have a `QVariantMap` in `QgsRenderContext::setCustomRenderFlags`, it would probably make sense to change the `QgsMapSettings` method as well. Unless they are all mapped to `keyname -> true` entries, where keynames are the ones listed in the `;`-separated string in `QgsMapSettings::mCustomRenderFlags`. As far as I'm concerned, it is suffiient to be able to test inclusion in the list, I don't need to pass other data.

[1] https://github.com/kadas-albireo/kadas-albireo2/blob/master/kadas/app/globe/kadasglobetilesource.cpp#L120
[2] https://github.com/kadas-albireo/kadas-albireo2/blob/master/kadas/app/mapgrid/kadasmapgridlayer.cpp#L213

Reply:
I'd suggest it would be cleaner to replace those with:

    settings.setCustomRenderFlag( "globe", true );

and

     bool drawLabels = !( renderFlags.customRenderFlag( "globe" ).toBool() || renderFlags.customRenderFlag( "kml" ).toBool() );

Using a QStringlist means that:

1. you're putting the responsibility on the caller to ensure they aren't blindly wiping away some useful existing flag, e.g. by calling `context.setCustomRenderFlags( ['myflag'])` when some flags already exist.

2. you're forcing all flags to go through a string value, so there's going to be loss of precision in some cases (e.g. trying to store a double value)

Reply:
I agree with you, but that would mean changing the API in `QgsMapSettings::setCustomRenderFlags` [1]. That probably wouldn't be acceptable for QGIS 3.x? Unless I mark that method as deprecated, add a new method

    QgsMapSettings::setCustomRenderFlags(const QString& key, bool active);

and replace the implementation of the old one with something like

    QgsMapSettings::setCustomRenderFlags(const QString& flags)
    {
        mCustomRenderFlags = QVariantList();
        const QStringList flaglist = flags.split(";");
        for(const QString& key : flaglist) {
            setCustomRenderFlags(key, true);
        }
    }
    

[1] https://github.com/qgis/QGIS/blob/master/src/core/qgsmapsettings.h#L224

Reply:
Oh I totally missed the QgsMapSettings API. Hmmm... My gut feeling would be to deprecate it and also use a variant map there, if you're ok with that change?